x-pack/plugins/observability_ai_assistant/server/functions/context.ts

```diff
@@ -77,49 +63,94 @@ export function registerRecallFunction({
         required: ['queries', 'categories'],
       } as const,
     },
-    async ({ arguments: { queries, categories }, messages, connectorId }, signal) => {
-      const systemMessage = messages.find((message) => message.message.role === MessageRole.System);
+    async ({ arguments: args, messages, connectorId, appContexts }, signal) => {
+      const { queries, categories } = args;
 
-      if (!systemMessage) {
-        throw new Error('No system message found');
-      }
+      async function getContext() {
+        const systemMessage = messages.find(
+          (message) => message.message.role === MessageRole.System
+        );
 
-      const userMessage = last(
-        messages.filter((message) => message.message.role === MessageRole.User)
-      );
+        const screenDescription = appContexts.map((context) => context.description).join('\n\n');
 
-      const nonEmptyQueries = compact(queries);
+        const content = { screen_description: screenDescription, learnings: [] };
```

Comment:
What about moving the `dataWithinTokenLimit` to the `getContext` function?

```suggestion
        // any data that falls within the token limit, send it automatically
        const dataWithinTokenLimit = compact(appContexts.flatMap((context) => context.data)).filter(
          (data) => encode(JSON.stringify(data.value)).length <= MAX_TOKEN_COUNT_FOR_DATA_ON_SCREEN
        );

        const content = {
          screen_description: screenDescription,
          data_on_screen: dataWithinTokenLimit,
          learnings: [],
        };
```